### PR TITLE
Implement udf_get_key_live

### PR DESCRIPTION
**Pull request by jklukas**
The existing implementation assumes key/value arrays are wrapped in a
single-field struct where the field is named key_value. That structure
is enforced by our Parquet import machinery, but the same is not true of the
GCP pipeline where map types are transformed into key/value arrays without
the extra layer of encapsulation.

This refactor makes this function general enough to handle either case.
The example in @sunahsuh's [documentation PR](https://github.com/mozilla/firefox-data-docs/pull/290)
would become slighlty more verbose, exposing the key_value field:

```
SELECT udf.get_key(event_map_values.key_value, 'reason') AS reason,
       COUNT(*) AS EVENTS
FROM telemetry.events
WHERE submission_date_s3 = DATE_SUB(CURRENT_DATE(), INTERVAL 1 DAY)
  AND event_category='normandy'
  AND event_method='unenroll'
GROUP BY 1
ORDER BY 2 DESC
```

### CODE REVIEW

**Comment by sunahsuh** [comment]
Is there any way we can fix this in the derived datasets instead, especially since we're planning on moving many of those into clustered tables soon anyway?

**Comment by relud** [comment]
> Is there any way we can fix this in the derived datasets instead

We could add a post-import transform for map types in https://github.com/mozilla/parquet2bigquery with a feature flag and update tables to not have the key_value sub-struct

**Comment by sunahsuh** [comment]
I should say explicitly: I think asking folks to remember where they have to unwrap a map and where they don't have to is very confusing, and those derived datasets are going to live on for a while longer so I very much think we should fix this before telling them they should start using it

**Comment by relud** [comment]
> I should say explicitly: I think asking folks to remember where they have to unwrap a map and where they don't have to is very confusing, and those derived datasets are going to live on for a while longer so I very much think we should fix this before telling them they should start using it

Fair, correction:

We ~~could~~ should add a post-import transform for map types

**Comment by jklukas** [comment]
We could solve this with a view on top of the imported datasets, too, that hides the sub-struct. That's likely easier than updating the parquet2bigquery code and coordinating a recreation of the table with deploying the updated logic.

**Comment by jklukas** [comment]
In particular, I think we could have a relatively simple script that walks tree of a table's json schema and outputs a relevant query with appropriate `* REPLACE (...)` statements to unnest any fields named `key_value`.

**Comment by sunahsuh** [comment]
@jklukas It'd probably look something like this 😁 https://github.com/mozilla/parquet2bigquery/pull/28/files

**Comment by jklukas** [comment]
> @jklukas It'd probably look something like this 😁 https://github.com/mozilla/parquet2bigquery/pull/28/files

Indeed! I hadn't had a chance yet to read through that code. If generating views seems preferable to altering the import mechanism, then you've indeed already written the bulk of the logic there.

Do you have an opinion on which approach would be preferable?

**Comment by jklukas** [comment]
It sounds like we think views are a good solution here, but that will likely take a while.

For the shorter term, what do you think about providing an additional `udf_get_key_parquet` that assumes we have the extra `key_value` layer. Folks would be able to use that on the existing main_summary_v4, then once we get the view in place, users would just replace uses of `udf_get_key_parquet` with `udf_get_key`.

**Comment by relud** [comment]
i'm going to try to make a udf that will use the get_key layer if it exists

**Comment by sunahsuh** [comment]
@jklukas 1. I think it's going to get pretty confusing for users to have to keep track of what's a parquet-based dataset and what's not and 2. if we do have two separate UDFs, I'd suggest it's the queries on live tables that should use a special udf (`udf_get_key_live`?) for now since there's far fewer users

@relud fwiw I spent a little time on this and ran into the problem of there not being any way to do schema/datatype introspection in SQL UDFs, since the input is maps of any type I wasn't able to make any assumptions about the output schema :/ Maybe you'll have better luck :)

**Comment by relud** [comment]
no luck (without using json extract, which isn't helpful because it limits the output to strings)

**Comment by jklukas** [comment]
> @jklukas 1. I think it's going to get pretty confusing for users to have to keep track of what's a parquet-based dataset and what's not and 2. if we do have two separate UDFs, I'd suggest it's the queries on live tables that should use a special udf (`udf_get_key_live`?) for now since there's far fewer users

Or maybe it's reasonable to expect users to use `udf_get_key` directly. When they need to transition to the less-nested format, they'd just be replacing `udf_get_key(some_field.key_value, 'mykey')` with `udf_get_key(some_field, 'mykey')`.

**Comment by relud** [comment]
> 2. if we do have two separate UDFs, I'd suggest it's the queries on live tables that should use a special udf (`udf_get_key_live`?) for now since there's far fewer users

what if we put the nested udf in derived-datasets and the unnested one in prod, and don't put datasets in prod until they are unnested (by view or import)?

**Comment by jklukas** [comment]
Compiling various thoughts, here's what I'd suggest:

- We leave `udf_get_key` as it is, assuming there's a key_value struct
- We introduce a `udf_get_key_live` temporarily, which doesn't expect a key_value struct
- We deploy both functions in both places
- Once we are prepped to present a view on top of main_summary that hides the key_value structs, - we send out an email alerting users to an impending breaking change, then we deploy that view along with dropping `udf_get_key_live` and changing `udf_get_key` to no longer reference `key_value`

**Comment by sunahsuh** [comment]
My only suggestion is:

> Once we are prepped to present a view on top of main_summary that hides the key_value structs, - we send out an email alerting users to an impending breaking change, then we deploy that view along with dropping udf_get_key_live and changing udf_get_key to no longer reference key_value

I don't think it has to be a breaking change, in that we should probably call the view on top of main summary `main_summary` (e.g. the user-facing name doesn't change) and we can change the udf at the same time as we do the switchover, so these would be user-invisible changes. `udf_get_key_live` would continue to work as-is (and, in fact, would basically be an alias for `udf_get_key`), but we can declare it deprecated.

File: udf/udf_get_key.sql [hunk 1]
```diff
@@ -1,3 +1,18 @@
+/*
+
+Fetch the value associated with a given key from a "map type" imported from
+a Parquet dataset.
+
+Our machinery for importing datasets from Parquet models map types as a field
+containing a repeated struct named "key_value" with fields key and value.
```
Comment (sunahsuh) [inline udf/udf_get_key.sql]: FWIW, I discovered in the course of looking into this that this is how parquet itself models the logical map type out of its primitive types: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps

**Comment by relud** [comment]
superceded by #239

**Pull request closed without merging**